tiledb/sm/c_api/tiledb.cc

```diff
@@ -2737,51 +1904,25 @@ int32_t tiledb_deserialize_query_and_array(
   if (query == nullptr) {
     return TILEDB_ERR;
   }
-
-  api::ensure_buffer_is_valid(buffer);
-
-  // Create array struct
-  *array = new (std::nothrow) tiledb_array_t;
-  if (*array == nullptr) {
-    auto st = Status_Error(
-        "Failed to create TileDB array object; Memory allocation error");
-    LOG_STATUS_NO_RETURN_VALUE(st);
-    save_error(ctx, st);
-    return TILEDB_OOM;
-  }
+  ensure_buffer_is_valid(buffer);
+  ensure_output_pointer_is_valid(array);
 
   // Check array URI
   auto uri = tiledb::sm::URI(array_uri);
   if (uri.is_invalid()) {
-    auto st = Status_Error("Failed to create TileDB array object; Invalid URI");
-    delete *array;
-    *array = nullptr;
-    LOG_STATUS_NO_RETURN_VALUE(st);
-    save_error(ctx, st);
-    return TILEDB_ERR;
+    throw CAPIException("Cannot deserialize query and array; Invalid URI");
   }
 
-  // Allocate an array object
-  try {
-    (*array)->array_ =
-        make_shared<tiledb::sm::Array>(HERE(), ctx->resources(), uri);
-  } catch (std::bad_alloc&) {
-    auto st = Status_Error(
-        "Failed to create TileDB array object; Memory allocation error");
-    delete array;
-    array = nullptr;
-    LOG_STATUS_NO_RETURN_VALUE(st);
-    save_error(ctx, st);
-    return TILEDB_OOM;
-  }
+  // Create array object
+  *array = tiledb_array_t::make_handle(ctx->resources(), uri);
```

Comment:
Need to account for possible resource leaks below this line. This handle could be created at the end of the function, once all the other sources of failure have been accounted for. Even there, though, because this code is returning two handles, the second one needs to be in a `try` block to break the first handle if the second one fails to construct.